DRY up Tomcat configuration and add support for Tomcat 8.5

boris-petrov opened the pull request:
DRY up Tomcat configuration and add support for Tomcat 8.5

By extracting a few new tomcatX projects that contain the common source
code between Tomcat versions 8, 8.5 and 9

Fixes #28
Fixes #30

javabrett commented on settings.gradle:
```diff
@@ -14,6 +14,7 @@ if (JavaVersion.current().isJava8Compatible()) {
 }
 include 'libs:gretty-runner-tomcat'
 include 'libs:gretty-runner-tomcat7'
+include 'libs:gretty-runner-tomcat8and9'
```
I don't suppose there's a way of doing this without locking that filename to Tomcat 8 and 9 - presumably we will need it for 10 also?  If Tomcat 7 is the odd-one-out, could we put this code into the base runner and make Tomcat 7 the exception?  Just 'supposin.

boris-petrov replied:
Well, I'm not sure. There is this `gretty-runner-tomcat` project that is compiled against Tomcat 7. Then, the rest of `gretty-runner-tomcat#NUMBER` projects depend on the "main" one and add additional stuff on top of it. If we made the "main" one include the things from `8and9`, then `7` would not compile as it doesn't have the correct dependencies. Not sure if I explain it good enough. Or do you have some other idea?

boris-petrov commented:
@javabrett, @henrik242 - I rewrote the commit to support Tomcat 8.5 also (sorry, things are too intertwined so I couldn't do it in a separate commit). Also, I've renamed the directories to be like `...-tomcat8+` which I guess give better meaning. This is as DRY as possible I think.

One thing I would also do (if you agree) is to rename the "master" `gretty-runner-tomcat` project to `gretty-runner-tomcat7+` to match the naming convention. What do you think?

javabrett commented:
Thanks @boris-petrov . Unfortunately the `+` is going to upset Maven coordinates:

> Invalid publication 'mavenJava': artifactId is not a valid Maven identifier ([A-Za-z0-9_\-.]+).

https://travis-ci.org/gretty-gradle-plugin/gretty/jobs/355729509#L1110

I don't currently have naming suggestions but I might think it over tomorrow.

boris-petrov commented:
@javabrett - ah, yes, I was thinking it might be a problem. Damn. `+` was a good name. We could use `8plus`? Or `8andAbove`? I don't like them but... tell me if you have any other ideas.

henrik242 commented:
+1 on 8plus.

boris-petrov commented:
@henrik242 - maybe you meant `plus1 on 8plus`? :D

henrik242 commented:
Obviously! :D

javabrett commented on libs/gretty-core/build.gradle:
```diff
@@ -29,6 +29,8 @@ tasks.processResources {
     tomcat7ServletApiVersion: project.tomcat7_servlet_api_version,
     tomcat8Version: project.tomcat8_version,
     tomcat8ServletApiVersion: project.tomcat8_servlet_api_version,
+	tomcat85Version: project.tomcat85_version,
```
Check indentation here, looks like a TAB.

javabrett commented:
Let me see if I understand this correctly.  Currently (master), we have the base `gretty-runner-tomcat` then the versioned `gretty-runner-tomcat7`, `gretty-runner-tomcat8`, `gretty-runner-tomcat9`.  Those version-specific libs currently only apply to the version they are named-for ... there's no inheritance/overrides 7 -> 8 -> 9, only (base) -> 7, (base) -> 8 etc.

This change would make it such that the `tomcat9` runner might inherit from `tomcat85` -> `tomcat8` ... as necessary, so for Tomcat 9 you might need all three libs.

In that case I'm not sure that we have a strict need for a rename to reflect that?  It seems reasonable that `gretty-runnter-tomcat` contains as much common code as practical, then any new or specific features/overrides appear in a specific-version library.  And that Tomcat 9, which enhances Tomcat 8.5, might have a smaller override library and also depend on Tomcat 8.5 etc.

I know this is not how it works now, where the versioning/inheritance is flat.

So - does it make sense to just not rename at all, and accept that Tomcat 10 might later have a `tomcat10` runner and all of the DRY overrides in 9, 8.5 etc.?

I've noticed that in a lot of cases identical classnames have been used.  That would need to be fixed?  Otherwise we might be relying on classloader order priority.

boris-petrov commented:
@javabrett - on `master` it is like this: (base) -> 7, (base) -> 8, 8 -> 9. This PR makes it like so: (base) -> 7, (base) -> 8+, 8+ -> 8, 8+ -> 85+, 85+ -> 85, 85+ -> 9. Now that I think about it, 85+ could be removed and 85 could be used instead of it. That would remove one of the projects. However, 8+ cannot be removed, as it is the base for both 8 and 85+, which actually have different `TomcatConfigurerImpl.groovy` implementations.

As for the last question - some of these classes are used from reflection (especially this `TomcatConfigurerImpl` - in `libs/gretty-runner-tomcat/src/main/groovy/org/akhikhl/gretty/ServerManagerFactory.groovy`) so that's why they have to have the same name. However, they are in different projects so I guess there is no classorder problem... I'm not sure though, it's still a bit black-magic to me what exactly happens here and when. :)

So:

1) Should I remove the redundant 85+?
2) What should we rename the `+` to?
3) Any other changes?

henrik242 commented:
1. Yes
2. 8plus
3. lgtm :)

javabrett commented:
That sounds reasonable - thanks.

boris-petrov commented:
I've made the changes and pushed. Not sure what is going on, but locally I'm getting:

```
Could not resolve all files for configuration ':integrationTests:buildSrc:gretty-integrationTest:compileClasspath'.
> Could not resolve de.undercouch:gradle-download-task:3.1.1.
  Required by:
      project :gretty-integrationTests-buildsrc:gretty-integrationTest
   > Could not resolve de.undercouch:gradle-download-task:3.1.1.
      > java.lang.UnsupportedClassVersionError: sun/security/rsa/RSACore
> Could not resolve org.gretty:gretty:2.1.1-SNAPSHOT.
  Required by:
      project :gretty-integrationTests-buildsrc:gretty-integrationTest
   > Could not resolve org.gretty:gretty:2.1.1-SNAPSHOT.
      > java.lang.UnsupportedClassVersionError: sun/security/rsa/RSACore has been compiled by a more recent version of the Java Runtime (class file version 18559.1542), this version of the Java Runtime only recognizes class file versions up to 52.0
```

When I try building... any ideas what's going on?

javabrett commented:
Looks great - thanks for persisting with this @boris-petrov .  Merging.
